spirv-opt: Switch from Vulkan.DebugInfo to Shader.DebugInfo

greg-lunarg opened the pull request:
Includes:
- Shift to use of spirv-header extinst.nonsemantic.shader grammar.json
- Remove extinst.nonsemantic.vulkan.debuginfo.100.grammar.json
- Enable all optimizations for Shader.DebugInfo

Also fixes scalar replacement to only insert DebugValue after all
OpVariables. This is not necessary for OpenCL.DebugInfo, but it is
for Shader.DebugInfo.

Likewise, fixes Private-to-Local to insert DebugDeclare after all
OpVariables.

Also fixes inlining to handle FunctionDefinition which can show up
after first block if early return processing happens.

greg-lunarg commented:
This PR is dependent on this PR: https://github.com/KhronosGroup/SPIRV-Headers/pull/236

jaebaek commented:
The bazel build fails because `source/extinst.nonsemantic.vulkan.debuginfo.100.grammar.json` file (see https://github.com/KhronosGroup/SPIRV-Tools/blob/master/build_defs.bzl#L44) does not exist.

I am not sure copying `extinst.nonsemantic.vulkan.debuginfo.100.grammar.json` file from https://github.com/KhronosGroup/SPIRV-Headers is allowed.
One option would be adding `spirv_ext_inst_nonsemantic_vulkan_debuginfo_100_grammar_unified1` to https://github.com/KhronosGroup/SPIRV-Headers/blob/master/BUILD.bazel and use it for `VKDEBUGINFO100_GRAMMAR_JSON_FILE` similar to `CLDEBUGINFO100_GRAMMAR_JSON_FILE` and `spirv_ext_inst_opencl_debuginfo_100_grammar_unified1` in https://github.com/KhronosGroup/SPIRV-Tools/blob/master/build_defs.bzl#L43

greg-lunarg commented:
Yes. That is my intent. My SPIRV-Headers fix has been merged, so hopefully it will make its way into CI shortly. If not, I might try updating known good to try and get the latest SPIRV-Headers into CI.

jaebaek commented on source/opt/inline_pass.cpp:
```diff
@@ -441,6 +441,11 @@ std::unique_ptr<BasicBlock> InlinePass::InlineBasicBlocks(
     auto tail_inst_itr = callee_block_itr->end();
     for (auto inst_itr = callee_block_itr->begin(); inst_itr != tail_inst_itr;
          ++inst_itr) {
+      // don't inline function definition links, the calling function is not a
```
`// Don't ..`

greg-lunarg replied:
Done

jaebaek commented on source/opt/local_access_chain_convert_pass.cpp:
```diff
@@ -333,6 +333,20 @@ bool LocalAccessChainConvertPass::AllExtensionsSupported() const {
     if (extensions_allowlist_.find(extName) == extensions_allowlist_.end())
       return false;
   }
```
same as the comment for ADCE

greg-lunarg replied:
Same response :)

jaebaek commented on source/opt/local_single_block_elim_pass.cpp:
```diff
@@ -188,6 +188,20 @@ bool LocalSingleBlockLoadStoreElimPass::AllExtensionsSupported() const {
     if (extensions_allowlist_.find(extName) == extensions_allowlist_.end())
       return false;
   }
```
likewise

greg-lunarg replied:
Same response :)

jaebaek commented on source/opt/local_single_store_elim_pass.cpp:
```diff
@@ -53,6 +53,20 @@ bool LocalSingleStoreElimPass::AllExtensionsSupported() const {
     if (extensions_allowlist_.find(extName) == extensions_allowlist_.end())
       return false;
   }
```
ditto

greg-lunarg replied:
Same response :)

jaebaek commented on source/opt/scalar_replacement_pass.cpp:
```diff
@@ -172,10 +172,14 @@ bool ScalarReplacementPass::ReplaceWholeDebugDeclare(
+    Instruction* insert_before = var->NextNode();
+    while (insert_before->opcode() == SpvOpVariable)
+      insert_before = insert_before->NextNode();
+    assert(insert_before != nullptr && "unexpected end of list");
```
Just out of curiosity.
Is there any reason to add `DebugValue` after all variables?
As far as I remember, the instruction layout allows us to put `DebugValue` between `OpVariable`s

greg-lunarg replied:
Because these are "nonsemantic" instructions, they are more restricted than the OpenCL.DebugInfo.100 instructions:

From https://github.com/KhronosGroup/SPIRV-Registry/blob/master/nonsemantic/NonSemantic.Shader.DebugInfo.100.asciidoc#binary-form 

DebugScope, DebugNoScope, DebugDeclare, DebugValue, DebugLine, DebugNoLine, and DebugFunctionDefinition instructions can interleave with the instructions within a function, but must appear within valid locations in a block as required by SPV_KHR_non_semantic_info. In particular this means **they cannot come before any OpPhi or function-level variable declarations in a block**, and they cannot come after a Merge Instruction.

jaebaek reviewed (commented):
Looks good to me with nits.

greg-lunarg commented:
The shaderc test is failing due to problems related to a merge done yesterday.

@jaebaek @alan-baker @dnovillo @dneto0 The ubuntu and macos bazel builds are failing due to a "missing" file, yet the windows bazel build is passing.

Does someone need to update "standing" files for these test platforms?

jaebaek commented:
Just change all

source/opt/instruction.h:35:#include "spirv/unified1/NonSemanticShaderDebugInfo100.h"
source/val/validate_layout.cpp:28:#include "spirv/unified1/NonSemanticShaderDebugInfo100.h"
source/val/validate_extensions.cpp:36:#include "spirv/unified1/NonSemanticShaderDebugInfo100.h"

to

```
#include "NonSemanticShaderDebugInfo100.h"
```

The new `generate_extinst_lang_headers()` in build_defs.bzl generates headers in that way.

greg-lunarg commented on source/opt/aggressive_dead_code_elim_pass.cpp:
```diff
@@ -145,6 +145,20 @@ bool AggressiveDCEPass::AllExtensionsSupported() const {
+    if (0 == std::strncmp(extension_name, "NonSemantic.", 12) &&
+        0 != std::strncmp(extension_name, "NonSemantic.Shader.DebugInfo.100",
+                          32)) {
+      return false;
```
Actually, OpenCL.DebugInfo.100 **is** allowed in ADCE, at least it is not disallowed here. The only thing that is being disallowed by this logic is any NonSemantic instruction sets that are **not** Shader.DebugInfo.100. This makes sense since we cannot safely assume in general that "dead" NonSemantic instructions are removable.

greg-lunarg replied:
Rather, we might want to ask ourselves: do we want to create an "allow list" of Extended Instruction Sets in the same way we have an "allow list" for Extensions? In the past this has not been necessary since before OpenCL.DebugInfo, extended instructions did not have side effects and "dead" instructions could be deleted. But now with OpenCL.DebugInfo and NonSemantic.Shader.DebugInfo we cannot make this assumption, and we need to add special logic to correctly handle these extended instruction sets.

I think we are currently handling all extended instruction sets correctly, so logic to disallow unhandled extended instruction sets is not immediately needed, but it is something we might want to consider to make it easier to add extended instruction sets in the future.

greg-lunarg commented on source/opt/aggressive_dead_code_elim_pass.cpp:
```diff
@@ -145,6 +145,20 @@ bool AggressiveDCEPass::AllExtensionsSupported() const {
     if (extensions_allowlist_.find(extName) == extensions_allowlist_.end())
       return false;
   }
```
Done

greg-lunarg commented:
I think the shaderc smoketest failure is not related to my changes. The errors look related to a merge that happened yesterday. I suspect the testing platform somehow needs to be updated. The shaderc smoketest had been passing with my changes and my recent changes should not have caused it to start failing.

greg-lunarg commented:
I still have some regression tests to add, so please hold off on merging until those are added.

jaebaek commented:
Sure

jaebaek commented:
The ubsan failure does not contain a log ..
Could you please fix the clang-format error? Let's check if it will report the ubsan failure again.

```
+ cd /tmpfs/src/github/SPIRV-Tools
+ git clone --depth=1 https://github.com/KhronosGroup/SPIRV-Headers external/spirv-headers
Cloning into 'external/spirv-headers'...
+ git clone https://github.com/google/googletest external/googletest
Cloning into 'external/googletest'...
+ cd external
+ cd googletest
+ git reset --hard 1fb1bb23bb8418dc73a5a9a82bbed31dc610fec7
HEAD is now at 1fb1bb23 Googletest export
+ cd ..
+ cd ..
+ git clone --depth=1 https://github.com/google/effcee external/effcee
Cloning into 'external/effcee'...
+ git clone --depth=1 https://github.com/google/re2 external/re2
Cloning into 'external/re2'...
+ curl -L --fail https://raw.githubusercontent.com/llvm/llvm-project/main/clang/tools/clang-format/clang-format-diff.py -o utils/clang-format-diff.py
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100  5510  100  5510    0     0  60642      0 --:--:-- --:--:-- --:--:-- 61222
++ date
+ echo Fri Sep 10 14:15:11 PDT 2021: Check formatting...
Fri Sep 10 14:15:11 PDT 2021: Check formatting...
+ ./utils/check_code_format.sh
Found formatting errors!
--- test/opt/local_ssa_elim_test.cpp	(before formatting)
+++ test/opt/local_ssa_elim_test.cpp	(after formatting)
@@ -2166,7 +2166,6 @@
 }

 TEST_F(LocalSSAElimTest, ShaderDebugForLoop) {
-
   const std::string text = R"(
 ; CHECK: [[f_name:%\w+]] = OpString "f"
 ; CHECK: [[i_name:%\w+]] = OpString "i"


[ID: 3904725] Build finished after 38 secs, exit value: 1
```

greg-lunarg commented:
You can see a log for the ubsan failure at https://github.com/KhronosGroup/SPIRV-Tools/pull/4520

greg-lunarg commented:
I believe that this can be checked in now. I will be removing the WIP designation.

You will see that I have added two NonSemantic.Shader.DebugInfo.100-specific tests. That might even be one more than we need.

The aggressive_dead_code_elim test I added shows that a spirv-opt pass can read the NonSemantic.Shader.DebugInfo.100 format. This is most significant for all the instructions here which differ from OpenCL.Shader.DebugInfo.100 in that they take OpConstant instead of literal operands. In addition, this test also contains DebugTypeBasic, DebugTypeMember and DebugFunction which have one less operand than their OpenCL counterpart. In addition, this test also contains DebugFunctionDefintion, which is unique to NonSemantic.Shader and makes sure that it is not eliminated. Finally, this tests demonstrates that the ADCE algorithm optimizes the NonSemantic.Shader form the same as the OpenCL form.

By induction from this test, all the other passes should optimize the NonSemantic.Shader format the same as the OpenCL format, at least for those operations which they have in common, which is essentially all that we are covering at this time. Thus, the OpenCL tests should be sufficient to cover NonSemantic.Shader format as well. As support for additional NonSemantic.Shader instructions is added (such as DebugLine and DebugNoLine, which are next), additional tests will be added.

The SSA test I added is just an extra sanity check that my assertion that NonSemantic.Shader will be optimized similar to OpenCL and thus is covered by all the OpenCL tests.

greg-lunarg commented:
BTW, I did fix the format error, but the ubsan failure still did not leave a log. Again, I would point out that both this PR and #4520 both have an ubsan failure which suggests the problem is not with these PRs. You can see a log for the ubsan failure at #4520.

jaebaek commented on test/opt/scalar_replacement_test.cpp:
```diff
@@ -1935,12 +1935,12 @@ OpName %6 "simple_struct"
 ; CHECK: [[repl0:%\w+]] = OpVariable %_ptr_Function_uint Function
+; CHECK: OpExtInst %void [[ext]] DebugValue [[dbg_local_var]] [[repl3]] [[deref_expr]] %int_3
+; CHECK: OpExtInst %void [[ext]] DebugValue [[dbg_local_var]] [[repl2]] [[deref_expr]] %int_2
+; CHECK: OpExtInst %void [[ext]] DebugValue [[dbg_local_var]] [[repl1]] [[deref_expr]] %int_1
```
NonSemantic.Shader.DebugInfo.100 [DebugDeclare](https://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/nonsemantic/NonSemantic.Shader.DebugInfo.100.html#DebugDeclare) has `Indexes` operands. Therefore, using `DebugDeclare` with `Indexes` operands would be better for the scalar replacement pass instead of `DebugValue` with `Deref` operation.

I guess you want to make it as a separate PR. Am I right?

jaebaek replied:
Oh .. it is OpenCL.DebugInfo.100. Never mind.

jaebaek commented:
Similar to my answer in #4520, I guess rebasing it will fix the UBSan failure.
